### PR TITLE
Crée des comptes mails et redirections pour les nouveaux membres

### PR DESCRIPTION
**Pull request by Morendil**

### CODE REVIEW

File: lib/betagouvbot/accountrequest.rb [hunk 1]
```diff
@@ -0,0 +1,90 @@
+# encoding: utf-8
+# frozen_string_literal: true
+
+module BetaGouvBot
```
Comment (maukoquiroga) [inline lib/betagouvbot/accountrequest.rb]: Doc 😄 ?

File: lib/betagouvbot/accountrequest.rb [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    def ovh
+      OVH::REST
+    end
+
+    def api
+      ovh.new(ENV['apiKey'], ENV['appSecret'], ENV['consumerKey'])
+    end
+  end
+
+  class RedirectAction
```
Comment (maukoquiroga) [inline lib/betagouvbot/accountrequest.rb]: Doc 😄 ?

File: lib/betagouvbot/accountrequest.rb [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    attr_accessor :name
+    attr_accessor :password
+
+    def initialize(name, password)
+      @name = name
+      @password = password
+    end
+
+    def execute
+      accounts = api
```
Comment (maukoquiroga) [inline lib/betagouvbot/accountrequest.rb]: Wouldn't it be better to memoize ? Ex : 

```ruby
def api
  @api ||= OVH::REST.new(ENV['apiKey'], ENV['appSecret'], ENV['consumerKey'])
end

def execute
  api
end
```

File: lib/betagouvbot/accountrequest.rb [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  class AccountAction
+    attr_accessor :name
+    attr_accessor :password
+
+    def initialize(name, password)
+      @name = name
+      @password = password
+    end
+
+    def execute
```
Comment (maukoquiroga) [inline lib/betagouvbot/accountrequest.rb]: If you define this as

```ruby
def call; end
```

you'll be able to call it as 

```ruby
AccountAction.new(foo, bar).call
AccountAction.new(foo, bar).()
# etc
```

File: lib/betagouvbot/accountrequest.rb [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+    class << self
+      def call(members, command)
+        member, redirect, password = command.split(' ')
+        members
+          .select { |author| author[:id] == member }
+          .flat_map { |author| request_account(author, redirect, password) }
+      end
+
+      def request_account(member, redirect_raw, password)
```
Comment (maukoquiroga) [inline lib/betagouvbot/accountrequest.rb]: 🐙

File: lib/betagouvbot/accountrequest.rb [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+      endpoint = '/email/domain/beta.gouv.fr/redirection'
+      existing = redirections.get(endpoint, from: address)
+      if existing.length >= 1
+        update = "#{endpoint}/#{existing[0]}/changeRedirection"
+        redirections.post(update, to: @redirect)
+      else
+        redirections.post(endpoint, from: address, to: @redirect, localCopy: 'false')
+      end
+    end
+
```
Comment (maukoquiroga) [inline lib/betagouvbot/accountrequest.rb]: private

File: lib/betagouvbot/accountrequest.rb [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    def initialize(name, redirect)
+      @name = name
+      @redirect = redirect
+    end
+
+    def execute
+      address = "#{@name}@beta.gouv.fr"
+      redirections = api
+      endpoint = '/email/domain/beta.gouv.fr/redirection'
```
Comment (maukoquiroga) [inline lib/betagouvbot/accountrequest.rb]: You can define this as a contant

File: lib/betagouvbot/accountrequest.rb [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+      existing = redirections.get(endpoint, from: address)
```
Comment (maukoquiroga) [inline lib/betagouvbot/accountrequest.rb]:
```ruby
ENDPOINT = '/email/domain/beta.gouv.fr/redirection'.freeze

def address
  @address ||= "#{@name}@beta.gouv.fr".freeze
end

def existing
  @existing ||= api.get(endpoint, from: address)
end
```

File: lib/betagouvbot/accountrequest.rb (hunk 4 above)
Comment (Morendil) [inline lib/betagouvbot/accountrequest.rb]: Thinking back on our IRL conversation earlier, I see a fundamental difference between `execute` (as used in Action objects) and `call` (as used in the top-level modules webhooks call into):
 - `call` takes parameters
 - `execute` is nullary (an Action is basically a method object)

It makes sense to me to keep the names distinct on this basis ?

**Comment by Morendil** [comment]
@maukoquiroga je merge, on poursuivra le refactoring sur `master`

File: lib/betagouvbot/accountrequest.rb (hunk 4 above)
Comment (maukoquiroga) [inline lib/betagouvbot/accountrequest.rb]: Personally, I see no difference:

```ruby
AcountAction.new(foo, bar).()
AcountAction.new(foo, bar).call
AcountAction.new(foo, bar).execute
AcountAction.new(foo, bar).perform
AcountAction.new(foo, bar).operate
#...

AcountAction.(foo, bar)
AcountAction.call(foo, bar)
AcountAction.execute(foo, bar)
AcountAction.perform(foo, bar)
AcountAction.operate(foo, bar)
```

**Pull request merged**